Source/WebCore/html/parser/HTMLDocumentParserFastPath.cpp

```diff
@@ -0,0 +1,1053 @@
+            return didFail(HTMLFastPathResult::FailedParsingCharacterReference);
+
+        if (reference[0] == '#') {
+            UChar32 res = 0;
+            if (reference[1] == 'x' || reference[1] == 'X') {
+                for (size_t i = 2; i < reference.size(); ++i) {
+                    Char c = reference[i];
+                    res *= 16;
+                    if (c >= '0' && c <= '9')
+                        res += c - '0';
+                    else if (c >= 'a' && c <= 'f')
+                        res += c - 'a' + 10;
+                    else if (c >= 'A' && c <= 'F')
+                        res += c - 'A' + 10;
+                    else
+                        return didFail(HTMLFastPathResult::FailedParsingCharacterReference);
+
+                    if (res > UCHAR_MAX_VALUE)
+                        return didFail(HTMLFastPathResult::FailedParsingCharacterReference);
+                }
+            } else {
+                for (size_t i = 1; i < reference.size(); ++i) {
+                    Char c = reference[i];
+                    res *= 10;
+                    if (c >= '0' && c <= '9')
+                        res += c - '0';
+                    else
+                        return didFail(HTMLFastPathResult::FailedParsingCharacterReference);
+
+                    if (res > UCHAR_MAX_VALUE)
+                        return didFail(HTMLFastPathResult::FailedParsingCharacterReference);
+                }
+            }
+            StringBuilder entity;
+            appendLegalEntityFor(res, entity);
+            for (unsigned i = 0; i < entity.length(); ++i)
+                out.append(entity[i]);
+            // Handle the most common named references.
+        } else if (reference == "amp")
+            out.append('&');
+        else if (reference == "lt")
+            out.append('<');
+        else if (reference == "gt")
+            out.append('>');
+        else if (reference == "nbsp")
+            out.append(0xa0);
+        else {
+            // This handles uncommon named references.
+            String inputString { reference.data(), static_cast<unsigned>(reference.size()) };
+            SegmentedString inputSegmented { inputString };
+            StringBuilder entity;
+            bool notEnoughCharacters = false;
+            if (!consumeHTMLEntity(inputSegmented, entity, notEnoughCharacters) || notEnoughCharacters)
+                return didFail(HTMLFastPathResult::FailedParsingCharacterReference);
+
+            for (unsigned i = 0; i < entity.length(); ++i)
+                out.append(entity[i]);
+        }
+    }
+
+    void didFail(HTMLFastPathResult result)
+    {
+        // This function may be called multiple times. Only record the result the first time it's called.
+        if (m_parsingFailed)
+            return;
+
+        m_parseResult = result;
+        m_parsingFailed = true;
+    }
+
+    template<class ResultType> ResultType didFail(HTMLFastPathResult result, ResultType res)
```

Comment:
These argument names: `result` and `res`, not very helpful. I suggest `returnValue` instead of `res`.

Reply:
Will use `ResultType` -> `ReturnValueType` for the template parameter as well, to avoid confusion.